README.md
Phone Classifier
================

Used to quickly classifiy a phone number into :landline, :mobile or :forbidden

Current Travis Build Status: [![Build Status](https://secure.travis-ci.org/mobino/phone_classifier.png?branch=master)](https://travis-ci.org/mobino/phone_classifier)

Installation
------------

    gem install phone_classifier

Usage
-----

    PhoneClassifier.new('41791234567').kind        # => :mobile
    PhoneClassifier.new('41791234567').mobile?     # => true
    PhoneClassifier.new('41791234567').forbidden?  # => false


Supported Countries
-------------------

* +1, USA (will only reject service numbers, all other numbers considered mobile). Blame the US numbering systems
* +30, Greece
* +31, Netherlands
* +32, Belgium
* +33, France
* +34, Spain
* +36, Hungary
* +39, Italy
* +40, Romania
* +41, Switerzland
* +43, Austria
* +44, UK
* +45, Denmark
* +46, Sweden
* +47, Norway
* +48, Poland
* +49, Germany
* +54, Argentina
* +60, Malaysia
* +61, Australia
* +65, Singapore
* +90, Turkey
* +94, Sri Lanka
* +91, India
* +98, Iran
* +212, Morocco
* +213, Algeria
* +233, Ghana
* +234, Nigeria
* +248, Seychelles
* +249, Sudan
* +255, Tanzania
* +350, Gibraltar
* +351, Portugal
* +352, Luxembourg
* +353, Ireland
* +354, Iceland
* +356, Malta
* +357, Cyprus
* +358, Finland
* +359, Bulgaria
* +370, Lithuania
* +372, Estonia
* +377, Monaco
* +386, Slovenia
* +420, Czech Republic
* +421, Slovakia
* +503, El Salvador
* +509, Haiti
* +852, Hong Kong
* +972, Israel


Where is it used
----------------

Used in [Mobino](http://mobino.com), a mobile payment service, to classifiy client phone numbers

Related
-------

PhoneClassifier uses the fantastic [Phony](http://github.com/floere/phony.git) library to normalize numbers. Phony is
the tool of choice if you need to work with phone numbers in Ruby.

Contributing
------------

1. Fork it
2. Create your feature branch
3. Commit your changes
4. Push to the branch
5. Create new Pull Request


history.txt
History
=======

0.3.3/0.3.4
-----------
* jfgroff: accept all [789]. blocks for India

0.3.2
-----
* jfgroff: added 075 mobile block for Switzerland

Years of missing history documentation here...
----------------------------------------------

0.0.11
------
* jcfischer & afussen: added Israel

0.0.10
------
* jcfischer: added Turkey, Australia

0.0.9
-----
* jcfischer: added Argentina, El Salvador

0.0.8
-----
* jcfischer: fix endless loop when trying to handle short numbers

0.0.7
-----
* jcfischer: Changed detection for Morocco

0.0.4
-----
* jcfischer: Added Haiti, Iran

0.0.3
-----
* jcfischer: Added Hong Kong

0.0.2
-----
* jcfischer: Added Malaysia and Singapore

0.0.1
-----
* jcfischer: Initial extraction and gemification from Mobino Admin

spec/spec_helper.rb
begin
  require 'bundler'
rescue LoadError => e
  require 'rubygems'
  require 'bundler'
end
Bundler.setup :test
Bundler.require

SPEC_ROOT = File.expand_path(File.dirname(__FILE__))

require File.expand_path '../../lib/phone_classifier', __FILE__

require 'spec_helper_extensions'
include SpecHelperExtensions

spec/spec_helper_extensions.rb
module SpecHelperExtensions

  # performance_of { do something here }.should < 0.0001
  #
  require 'benchmark'
  # Try to load the Phony::PerformanceRatio from
  # the file spec/performance_ratio.rb - set it to the
  # ratio you think it is compared to my MacBook 2.66GHz i7.
  #
  # Example:
  #   Phony::PerformanceRatio = 0.8 # for 80% of my MacBook speed.
  #
  begin
    require File.expand_path '../performance_ratio', __FILE__
  rescue LoadError => e
    Phony::PerformanceRatio = 0 # Ignore speed tests by default.
  end
  def performance_of &block
    GC.disable
    result = Benchmark.realtime &block
    GC.enable
    result * Phony::PerformanceRatio
  end

  # Allows the definition of "describe-wide" examples and describes.
  #
  # Case for "it". Use in the describe block
  #   shared :it_should_behave_like_having_a_labels_class do
  #     response.should have_tag('.labels')
  #   end
  # Then, call it in the describe block or nested describe blocks as such
  #   it_should_behave_like_having_a_labels_class
  #
  # Case for "describe". Use in the describe block
  #   shared :describe_an_update_stream_item do
  #     before(:each) do
  #       # ...
  #     end
  #     it "bla" do
  #       response.should have_tag('.labels')
  #     end
  #   end
  # Then, call it in the describe block or nested describe blocks as such
  #   describe_an_update_stream_item
  #
  def shared(name, &block)
    self.class.class_eval do
      define_method name do
        name_ary = name.to_s.split('_')
        method_name = name_ary.first.to_sym
        descr_name  = name_ary[1..-1].join(" ")
        if [:describe, :it].include? method_name
          self.send(method_name, descr_name, &block)
        else
          raise ArgumentError, "Shared it or describe needs to start with 'it_' or 'describe_'."
        end
      end
    end
  end

  # Used to test private methods.
  #
  # The idea is to replace
  #   instance.send :private_method.
  # with
  #   in_the instance do
  #     private_method
  #   end
  #
  # Note: Now it is rather like "We have the scenario that we are in the given instance."
  #
  def in_the(instance, &block)
    instance.instance_eval(&block)
  end

end

lib/phone_classifier/forbidden.rb
module Forbidden

  def forbidden?
    cc, *national_parts = Phony.split Phony.normalize(@number)
    ndcs = Forbidden.data[cc]

    # if we don't know the country, we can make no assumptions
    return false unless ndcs
    ndc_forbidden_check(national_parts, ndcs)
  end


  private

  def ndc_forbidden_check(national_parts, ndcs)
    prefix = national_parts.shift

    # countries without NDCs (Denmark) have this parts structure ["45", false, "40", "53", "25", "77"]
    until national_parts.size == 0 || prefix.is_a?(String)
      prefix = national_parts.shift
    end

    # without a valid prefix, there is no check that we can make here
    return false unless prefix.is_a?(String)

    ndcs.each { |n| return true if prefix.match(/^#{n}$/) }
    false
  end

  def self.data
    @@forbidden ||= {
        '1' => %w{ 1.. }, # USA
        '30' => %w{ 8.. 9.. }, # Greece
        '31' => %w{84 85 87 88 91 112 676 800 900 906 909}, # Netherlands
        '32' => %w{70 800 90[0-9]}, # Belgium
        '33' => %w{8}, # France
        '34' => %w{10.. 80. 90.}, # Spain
        '36' => %w{ 40 41 50 55 60 71 80 90 91 }, # Hungary
        '39' => %w{1 7 8}, # Italy
        '40' => %w{ 8. 9. }, # Romania
        '41' => %w{ 74 800 840 842 844 848 860 868 900 901 906 }, # Switzerland
        '43' => %w{ 718 720 780 8.. 9.. }, # Austria
        '44' => %w{800 808 842 843 844 845 870 871 872 873 900 908 909 982}, # UK
        '45' => %w{10 11 18 70 80 90}, # Denmark
        '46' => %w{ 900 939 944 99. },
        '47' => %w{ 8.. }, # Norway
        '48' => %w{70., 80.}, # Poland
        '49' => %w{ 1. 164 168 169 180 181 19[0-4] 800 900 }, # Germany
        '52' => %w{ }, # Mexico
        '54' => %w{ 6.. 8.. }, # Argentina
        '55' => %w{ }, # Brazil
        '60' => %w{ 100 101 102 103 104 108 112 991 994 995 999 }, # Malaysia
        '61' => %w{ 1 }, # Australia
        '65' => %w{ 3... 800 1800 1900 }, # Singapore
        '90' => %w{ 592 594 596 522 900 800 850 822 512 }, # Turkey
        '91' => %w{ 90 95 96 }, # India
        '94' => %w{ }, # Sri Lanka
        '98' => %w{ 95. 96. 97. 98. 99.} , #Iran
        '212' => %w{ 8. }, # Morocco
        '213' => %w{ 8.. 9.. }, # Algeria
        '233' => %w{ 1 4 6 7 8 9 }, # Ghana
        '234' => %w{ 8.. 9.. }, # Nigeria
        '243' => %w{  }, # D.R. Congo
        '248' => %w{ 5 6 7 8 9 }, # Seychelles
        '249' => %w{ }, # Sudan
        '255' => %w{ 8.. 9.. }, # Tanzania
        '260' => %w{ }, # Zambia
        '350' => %w{ 8... }, # Gibraltar
        '351' => %w{ 800 80. }, # Portugal
        '352' => %w{ 118 12 13 800 801 900 901 905 }, # Luxembourg
        '353' => %w{ 76 80 81 800 }, # Ireland
        '354' => %w{ 8.. 9.. }, # Iceland
        '356' => %w{ }, # Malta
        '357' => %w{ 80 90 }, # Cyprus
        '358' => %w{ 6.. 7.. 8.. }, # Finland
        '359' => %w{ 800 900 }, # Bulgaria
        '370' => %w{ 800 9.. }, # Lithuania
        '372' => %w{ 1.. 800 }, # Estonia
        '375' => %w{  }, # Belarus
        '377' => %w{  }, # Monaco
        '386' => %w{ 80 89 90 }, # Slovenia
        '420' => %w{ 8.. 9.. }, # Czech republic
        '421' => %w{ }, # Slovakia
        '503' => %w{ }, # El Salvador
        '509' => %w{ 8. 9. }, # Haiti
        '852' => %w{ 7... 8...}, # Hong Kong
        '961' => %w{  }, # Lebanon
        '964' => %w{ 0 8 9 }, # Iraq
        '970' => %w{ 1 }, # Israel (blocked cc)
        '972' => %w{ 1 }, # Israel
    }
  end
end


lib/phone_classifier/mobile.rb
module Mobile

  def mobile?
    cc, *national_parts = Phony.split Phony.normalize(@number)
    mobile_fn = Mobile.data[cc]

    # if we don't know the country, we can make no assumptions
    return false unless mobile_fn
    mobile_fn.(national_parts)
  end


  private

  # Checks the NDC with the given list to see if it corresponds to a mobile device
  #
  # TODO 2014-11-21/tito: we need to merge this code with the regex checking code. I didn't want to make
  # too many changes when adding Zambia and Brazil, this is why only those two countries are currently running
  # through the other method. This method is a copy of is_number_of_type found previously in the PhoneClassifier
  # with one or two bug fixes (check for prefix being a String)
  def self.mobile_via_ndc?(national_parts, ndcs)
    prefix = national_parts.shift

    # countries without NDCs (Denmark) have this parts structure ["45", false, "40", "53", "25", "77"]
    until national_parts.size == 0 || prefix.is_a?(String)
      prefix = national_parts.shift
    end

    # without a valid prefix, there is no check that we can make here
    return false unless prefix.is_a?(String)

    ndcs.each { |n| return true if prefix.match(/^#{n}$/) } unless ndcs.nil?
    false
  end

  # Uses the given regex to check the national number (national destination code + subscriber #) for
  # a mobile device.
  def self.mobile_via_national?(national_parts, regex)
    national = national_parts.join('')
    national.match(regex) != nil
  end

  def self.data
    @@mobile ||= {
        '1'  => ->(x) { mobile_via_ndc?(x, %w{ [2-9].. }) }, # USA
        '27' => ->(x) { mobile_via_ndc?(x, %w{ 6[0-5] 7[0-4] 7[6-9] 8[1-4] }) }, # South Africa
        '30' => ->(x) { mobile_via_ndc?(x, %w{ 6.. }) }, # Greece
        '31' => ->(x) { mobile_via_ndc?(x, %w{ 6 }) }, # Netherlands
        '32' => ->(x) { mobile_via_ndc?(x, %w{4..}) }, # Belgium
        '33' => ->(x) { mobile_via_ndc?(x, %w{6 7}) }, # France
        '34' => ->(x) { mobile_via_ndc?(x, %w{6.. 7[1-9].}) }, # Spain
        '36' => ->(x) { mobile_via_ndc?(x, %w{ 20 30 31 70 }) }, # Hungary
        '39' => ->(x) { mobile_via_ndc?(x, %w{3..}) }, # Italy
        '40' => ->(x) { mobile_via_ndc?(x, %w{ 7. }) }, # Romania
        '41' => ->(x) { mobile_via_ndc?(x, %w{ 7[5-9] }) }, # Switzerland
        '43' => ->(x) { mobile_via_ndc?(x, %w{ 67 68 644 65. 66. 67. 68. 69. }) }, # Austria
        '44' => ->(x) { mobile_via_ndc?(x, %w{ 7[4-9].. }) }, # UK
        '45' => ->(x) { mobile_via_ndc?(x, %w{ 2[0-9] 30 31 40 41 42 5[0-3] 60 61 71 81 9[1-3] }) }, # Denmark
        '46' => ->(x) { mobile_via_ndc?(x, %w{ 70 72 73 76 }) }, # Sweden
        '47' => ->(x) { mobile_via_ndc?(x, %w{ 9.. 4.. }) }, # Norway
        '48' => ->(x) { mobile_via_ndc?(x, %w{50. 51. 53. 60. 66. 69. 72. 78. 79. 88.}) }, # Poland
        '49' => ->(x) { mobile_via_ndc?(x, %w{ 151 152.? 157.? 159 160 162 163 1672 17[0-9] }) }, # Germany
        '52' => ->(x) { mobile_via_national?(x, /^1\d/) }, # Mexico
        '54' => ->(x) { mobile_via_ndc?(x, %w{ \d{2} \d{3} \d{4} }) }, # Argentina (Everything can be mobile)
        '55' => ->(x) { mobile_via_national?(x, /^(11|12|13|14|15|16|17|18|19|21|22|24|27|28|91|92|93|94|95|96|97|98|99)9|^(31|32|33|34|35|37|38|41|42|43|44|45|46|47|48|49|51|52|53|54|55|61|62|63|64|65|66|67|68|69|71|73|74|75|77|79|81|82|83|84|85|86|87|88|89)[6-9]/)}, # Brazil
        '60' => ->(x) { mobile_via_ndc?(x, %w{ 10 11 12 13 14 153 154 156 158 16 17 18 19 }) }, # Malaysia
        '61' => ->(x) { mobile_via_ndc?(x, %w{ 4.. }) }, # Australia
        '65' => ->(x) { mobile_via_ndc?(x, %w{ 8... 9... }) }, # Singapore
        '90' => ->(x) { mobile_via_ndc?(x, %w{ 53. 54. 50. 55. }) }, # Turkey
        '91' => ->(x) { mobile_via_ndc?(x, %w{ [789]. }) }, # India
        '94' => ->(x) { mobile_via_ndc?(x, %w{ 71 72 75 77 78 }) }, # Sri Lanka
        '95' => ->(x) { mobile_via_ndc?(x, %w{ 9 }) }, # Myanmar
        '98' => ->(x) { mobile_via_national?(x, /^(91|92|93)|^(901|902)/) } , #Iran
        '212' => ->(x) { mobile_via_national?(x, /^6/) }, # Morocco
        '213' => ->(x) { mobile_via_ndc?(x, %w{ 55 66 697 699 77 790 796 }) }, # Algeria
        '225' => ->(x) { mobile_via_national?(x, /^(01|02|03|04|05|06|07|08|09|44|45|46|47|48|49|50|54|60|66|67|69|77|78)/) }, # Côte d'Ivoire
        '233' => ->(x) { mobile_via_ndc?(x, %w{ 20 23 24 26 27 28 50 54 57 }) }, # Ghana
        '234' => ->(x) { mobile_via_national?(x, /^(70|80|81|90)[0-9]/) }, # Nigeria
        '242' => ->(x) { mobile_via_ndc?(x, %w{ 01 04 05 06 }) }, #Congo
        '243' => ->(x) { mobile_via_ndc?(x, %w{ 80. 81. 82. 84. 88. 89. 97. 98. 99. }) }, #D.R. Congo
        '248' => ->(x) { mobile_via_ndc?(x, %w{ 2 }) }, # Seychelles
        '249' => ->(x) { mobile_via_ndc?(x, %w{ 9.  }) }, # Sudan
        '254' => ->(x) { mobile_via_ndc?(x, %w{ 7.. }) }, # Kenya
        '255' => ->(x) { mobile_via_ndc?(x, %w{ 61 65 68 7. }) }, # Tanzania
        '260' => ->(x) { mobile_via_national?(x, /^(955)|(966)|(97[7-9])/) },  # Zambia
        '263' => ->(x) { mobile_via_ndc?(x, %w(71 73 77 78)) }, # Zimbabwe
        '350' => ->(x) { mobile_via_ndc?(x, %w{ 5. 6. }) }, # Gibraltar
        '351' => ->(x) { mobile_via_ndc?(x, %w{ 9. }) }, # Portugal
        '352' => ->(x) { mobile_via_ndc?(x, %w{ 621 628 661 668 691 698 }) }, # Luxembourg
        '353' => ->(x) { mobile_via_ndc?(x, %w{ 8[2-9] }) }, # Ireland
        '354' => ->(x) { mobile_via_ndc?(x, %w{ 6.. 7.. 8..}) }, # Iceland
        '356' => ->(x) { mobile_via_ndc?(x, %w{ 77 79 98 99 }) }, # Malta
        '357' => ->(x) { mobile_via_ndc?(x, %w{ 95 96 97 99 }) }, # Cyprus
        '358' => ->(x) { mobile_via_ndc?(x, %w{ 4. 50 }) }, # Finland
        '359' => ->(x) { mobile_via_ndc?(x, %w{ 430 87. 88. 89. 98. }) }, # Bulgaria
        '370' => ->(x) { mobile_via_ndc?(x, %w{ 6.. }) }, # Lithuania
        '372' => ->(x) { mobile_via_ndc?(x, %w{ 5...? 81. 82.}) }, # Estonia
        '375' => ->(x) { mobile_via_ndc?(x, %w{ 25 29 33 44}) }, # Belarus
        '377' => ->(x) { mobile_via_ndc?(x, %w{ 4. 6 }) }, # Monaco
        '386' => ->(x) { mobile_via_ndc?(x, %w{ 30 40 31 41 51 71 70 64}) }, # Slovenia
        '420' => ->(x) { mobile_via_ndc?(x, %w{ 60[1-8] 72. 73. 77. 79. 91 }) }, # Czech Republic
        '421' => ->(x) { mobile_via_ndc?(x, %w{ 9.. }) }, # Slovakia
        '503' => ->(x) { mobile_via_ndc?(x, %w{ 6... 7... }) }, # El Salvador
        '509' => ->(x) { mobile_via_ndc?(x, %w{ 32 33 34 35 36 37 38 39 4. }) }, # Haiti
        '852' => ->(x) { mobile_via_ndc?(x, %w{ 5... 6... 9... }) }, # Hong Kong
        '961' => ->(x) { mobile_via_ndc?(x, %w{ 3. 7. }) }, # Lebanon
        '964' => ->(x) { mobile_via_ndc?(x, %w{ 7[1-9] }) }, # Iraq
        '970' => ->(x) { mobile_via_ndc?(x, %w{ 5. }) }, # Israel (blocked cc)
        '972' => ->(x) { mobile_via_ndc?(x, %w{ 5. }) }, # Israel
        '998' => ->(x) { mobile_via_national?(x, /^9[01234789]/) }, # Uzbekistan
    }
  end
end
